[Tooling] Use `buildkite_pipeline_upload` action to inline Buildkite release pipelines

iangmaia opened the pull request:
## Description

iOS counterpart: https://github.com/woocommerce/woocommerce-android/pull/12839
Related: pdccJN-xR-p2

This PR uses the `release-toolkit` action `buildkite_pipeline_upload` implemented on https://github.com/wordpress-mobile/release-toolkit/pull/597.

Our release pipelines (such as `complete-code-freeze`, `finalize-release`, etc) usually require to start a separate pipeline to make a release build.
The changes in this PR will result in add the steps of a release pipeline to the currently running build, without starting a new build altogether.

I've also added a trigger to run Danger whenever a PR moves to the "Ready For Review" state, as previously we relied on this behavior from Buildkite, but now it has changed.

## Testing 
I've added (and later reverted) a test commit ebf9a03b2f8b3c0417533531369f2efa0860c9de to test the `complete_code_freeze` lane.

These are resulting test builds created using the Buildkite Web UI directly with `PIPELINE=release-pipelines/complete-code-freeze.yml`:
- https://buildkite.com/automattic/woocommerce-android/builds/24827
- https://buildkite.com/automattic/woocommerce-android/builds/24828 (with `INCLUDE_WEAR_APP=true` set on Buildkite)

This PR can be tested in a similar way to the test above: creating a branch off this one, slightly modifying a release lane to prevent side effects and running the build directly on Buildkite.

iangmaia committed:
Update release-toolkit

iangmaia committed:
Use `buildkite_pipeline_upload` action

iangmaia committed:
Add `ready_for_review` GHA trigger so that Danger runs when a PR moves to ready for review

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iangmaia committed:
[TO REVERT] Disable side effects to complete-code-freeze pipeline for testing

iangmaia committed:
Revert "[TO REVERT] Disable side effects to complete-code-freeze pipeline for testing"

This reverts commit 2c3862fb4b73d3ccec88f1b689efe31130a4ad33.

spencertransier reviewed (approved):
:shipit:

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -701,27 +705,36 @@ platform :android do
+    if is_ci
+      buildkite_pipeline_upload(
+        pipeline_file: BUILDKITE_RELEASE_PIPELINE_FILE_PATH,
+        env_file: BUILDKITE_SHARED_PIPELINE_VARS_FILE_PATH,
```
Didn't we decide to make this parameter already default to `.buildkite/shared-pipeline-vars`, so that we wouldn't have to type its whole path and pass it explicitly on every call of `buildkite_pipeline_upload` we'd add to all our various client repos?

Which would allow us to not only remove this extra attribute, but also remove the `BUILDKITE_SHARED_PIPELINE_VARS_FILE_PATH` constant from line 39… and would also probably allow us to remove `BUILDKITE_CONFIG_DIR_PATH` , and use `BUILDKITE_RELEASE_PIPELINE_FILE_PATH = File.join('.buildkite', BUILDKITE_RELEASE_PIPELINE)` directly on line 41… simplifying the integration of this use of `buildkite_pipeline_upload` action not only in this PR but also in all the other clients in which we'll want to do a similar integration…

mokagio replied:
FWIW, that's what I ended up with in Simplenote iOS yesterday 

https://github.com/Automattic/simplenote-ios/blob/83fc950a42802b9f2e07916c2de6f3b9cc67f83c/fastlane/lanes/release.rb#L330-L353

So that's an example of how it might look like here.

iangmaia replied:
Ahh yes indeed. I copied this setup from DOAndroid, which wasn't updated to rely in the default parameter. I'll update it here and in DOAndroid, thanks!

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -35,6 +35,10 @@ METADATA_SOURCE_CHANGELOG_FILE_PATH = {
   WEAR_APP => File.join(METADATA_SOURCE_DIR_PATH[WEAR_APP], 'changelogs', 'default.txt')
 }.freeze
 DEV_RELEASE_NOTES_PATH = File.join(PROJECT_ROOT_FOLDER, 'RELEASE-NOTES.txt')
+BUILDKITE_CONFIG_DIR_PATH = File.join(PROJECT_ROOT_FOLDER, '.buildkite')
```
Not convinced it's worth it for this path to be absolute (and thus looking very long in call logs), while given how it's used, and paths passed to fastlane actions are relative to the repo root (not to the `fastlane/` subfolder), a relative path should work just as well, while being simpler?
```suggestion
BUILDKITE_CONFIG_DIR_PATH = '.buildkite'
```

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -701,27 +705,36 @@ platform :android do
+    # If we're running on CI, we can directly start the release pipeline jobs within the same build
+    if is_ci
+      buildkite_pipeline_upload(
+        pipeline_file: BUILDKITE_RELEASE_PIPELINE_FILE_PATH,
```
I actually think we could consider making the `pipeline_file:` argument of the `buildkite_pipeline_upload` action in release-toolkit automatically prepend `.buildkite/` to the path provided to its parameter if that value is not an absolute path…

Sure, it would require a new release of `release-toolkit` to add that small change, but that will further help simplify the call site of that action in the `Fastfile` of all the repos we plan to adopt this in the future, as we'd then be able to avoid having to add constants like `BUILDKITE_RELEASE_PIPELINE_FILE_PATH` and instead be able to just pass the `BUILDKITE_RELEASE_PIPELINE` constant already used for `buildkite_trigger_build`—without having to duplicate each of the constant to prepend the `.buildkite/` prefix when calling `buildkite_pipeline_upload` instead of `buildkite_trigger_build` (here there's only one, but I bet for some client repos there might be more, e.g. `beta-release.yml` vs `alpha-release.yml` vs `final-release.yml`?)

mokagio replied:
My two cents, I appreciate the option of having the pipeline file anywhere in the project. Buildkite [allows](https://buildkite.com/docs/agent/v3/cli-pipeline#uploading-pipelines-description) for the pipeline to come from anywhere, with reasonable defaults.

The Buildkite actions we've built in `release-toolkit` are agnostic of our projects setup and I like that. It makes me feel like other people could adopt them, too.

Of course, we should optimize to make our devs workflow good first, but in this instance specifying the full path seems like a good tradeoff.

AliSoftware replied:
> The Buildkite actions we've built in release-toolkit are agnostic of our projects setup and I like that. It makes me feel like other people could adopt them, too.

That's not really true though 😅

For example `buildkite_trigger_build` have the hardcoded assumption that the pipeline we want to run will be provided by the env var named `PIPELINE` and that its value is a path relative to `.buildkite/`

But this action only works because we have our convention in our own projects of having the root step of all our pipelines be written as `buildkite-agent pipeline upload ".buildkite/${PIPELINE:-pipeline.yml}"` and that's the assumption our `buildkite_trigger_build` action is entirely based on to allow triggering a specific `pipeline_file:`, which has nothing to do with Buildkite's internal default behavior, yet we've chosen to hardcode those defaults in our actions anyway 🙃

Proof: the call site of the `buildkite_trigger_build(…)` action a couple of lines below this one uses `pipeline_file: 'release-build.yml'` as that action's parameter, not `pipeline_file: '.buildkite/release-build.yml'`

mokagio replied:
The we might as well keep being selfish devils and optimize just for ourselves 

![image](https://github.com/user-attachments/assets/aecacd45-2621-4cc8-a32c-f4c04ca8a615)

iangmaia replied:
Yeah, I have mixed feelings about this. It's a similar issue I had with the `shared-pipeline-vars` when implementing the original action, though in that case at least it's just a default that's easy to override 🤔 

I get that some actions might rely too much in our setup, but shouldn't we instead strive for things to be more flexible instead of less (specially when the benefit isn't that big)?
It's also a bit about the work of going back, change that again, release a new version, change the apps (though admittedly, in this case only DOAndroid and Simplenote are using the new action so far, and the difference between the new one and the old `buildkite_trigger_build` is annoying to see) etc, for little benefit.

Though if we all agree there's no point in catering for that and focusing in our conventions (which indeed makes sense) I don't mind it.

iangmaia replied:
While we're here (considering a new `release-toolkit` version), we might as well already improve the `environment` parameter handling.

I've created this PR with both changes for us to consider: https://github.com/wordpress-mobile/release-toolkit/pull/608

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -701,27 +705,36 @@ platform :android do
-    )
+    # If we're running on CI, we can directly start the release pipeline jobs within the same build
+    if is_ci
+      buildkite_pipeline_upload(
```
Note for other repos who might use this PR for inspiration:

👍  Looking at the `Fastfile` of this repo, I see that WCAndroid have separate `start_code_freeze` and [`complete_code_freeze`, which means that the first beta that will be build—via a call to this `trigger_release_build` lane at the end of code-freeze](https://github.com/woocommerce/woocommerce-android/blob/trunk/fastlane/Fastfile#L222-L237)—will already be on a CI build that was triggered from the `release/*` branch, because the `complete_code_freeze` lane that will call this `trigger_release_build` will itself already have been called on the already-created `release/*` branch at that time. So everything should work out as expected.

🎗️ That being said, if someone were to use this PR as inspiration to similarly integrate the use of `buildkite_pipeline_upload` in the `Fastfile` of other client repos—especially ones in which the code freeze is not broken in 2 tasks (`start_code_freeze` and `complete_code_freeze`) but instead use a single `code_freeze` lane doing all in once task—then one needs to remember that for these other kind of setups, we'll have to add a `git checkout release/$VERSION` call at the start of the command script doing the beta build, because in those cases the `code_freeze` lane is run from the default branch of those repos—since that lane is the one creating the future `release/*` branch in the first place—and if we `buildkite_pipeline_upload` on such a CI build it will run the steps added from that pipeline on the same default branch of that CI build too instead of running the beta build on the newly-created `release/*` branch.

iangmaia replied:
Good point 👍

iangmaia committed:
Apply suggestions from code review

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

iangmaia committed:
Remove `env_file` parameter to rely on the default value

iangmaia committed:
Update release-toolkit

iangmaia committed:
Use only pipeline file name as the prefix `.buildkite/` will be added by `release-toolkit`

iangmaia committed:
Add `edited` GHA trigger so that Danger runs when screenshots are added to a description

iangmaia committed:
Forward RELEASE_VERSION to release builds and add it as parameter to checkout script

iangmaia committed:
Fix workflow reference

iangmaia commented:
@AliSoftware @mokagio  FYI, I've added the commit 7fef212f78f884d89e0430d8bf4b71b53a28ebdf to handle the situation highlighted by @mokagio [here](https://github.com/woocommerce/woocommerce-ios/pull/14244#pullrequestreview-2406797366), so now the release pipeline should also checkout the release branch.
While at it, I've changed the checkout script to receive the `RELEASE_VERSION` an argument as we do on iOS, for consistency.

mokagio commented on .buildkite/commands/checkout-release-branch.sh:
```diff
@@ -1,11 +1,6 @@
-    echo "RELEASE_VERSION is not set."
-    exit 1
-fi
+RELEASE_VERSION="${1?Please provide a release version as an argument.}"
```
TIL about the `${parameter?word}` expansion with `set -u` combo.

From `bash` manpage

> When not performing substring expansion, using the forms documented below (e.g., :-), bash tests for a parameter that is unset or null.  Omitting the colon results in a test only for a parameter that is unset.
>
> [...]
>
> `${parameter:?word}` Display Error if Null or Unset.  If parameter is null or unset, the expansion of word (or a message to that effect if word is not present) is written to the standard error and the shell, if it is not interactive, exits.  Otherwise, the value of parameter is substituted.

Nice

mokagio replied:
Nitpick: I find it DRY to move the Buildkite header in the script:
```suggestion
RELEASE_VERSION="${1?Please provide a release version as an argument.}"

echo '--- :git: Checkout Release Branch'

```

AliSoftware replied:
> TIL about the `${parameter?word}` expansion with `set -u` combo.
> 
> From `bash` manpage

My favorite cheat sheet I use as reference for those kind of things: https://devhints.io/bash

mokagio commented on .buildkite/release-builds.yml:
```diff
@@ -10,6 +10,9 @@ agents:
 steps:
   - label: "Gradle Wrapper Validation"
     command: |
+      echo '--- :git: Checkout Release Branch'
+      .buildkite/commands/checkout-release-branch.sh "$RELEASE_VERSION"
+
```
Followup to the nitpick comment above, the reason I like it is that every call simplifies to

```suggestion
      .buildkite/commands/checkout-release-branch.sh "$RELEASE_VERSION"

```

Which I'll admit is not much, but ensures that if we change the string, it'll changes consistently across all usages.

iangmaia committed:
Move echo call from pipelines to checkout-release-branch.sh script
